### PR TITLE
[DPE-3054] PeerRelation withing the same guidelines as Requires and Provides

### PR DESCRIPTION
**Pull request by juditnovak**
Implementing Peer Relations data abstraction and data interaction as the exact same like what we have for Cross-charm relations.

The proposal is FULLY re-using already existing code (behavior) set for Cross-charm Relations.

There are 3 Peer-specific requirements (how it's been implemented in the charms) that are taken over here:

Rolling upgrades
 1. If only databag was used (i.e. Internal Secrets in plain text) -> We move secrets one-by-one to a Juju Secret whenever the secret field is referred to
 2. If the charm was already using secrets BUT not by labels but URI ->  We remove the URI from the databag and stick a label on the secret
 
 Protection against old bugs (i.e. versions compatibilty):
 
 3. "deleted label": Soft delete, i.e. if a secret is to be removed, we just replace the value with a specific label

### CODE REVIEW

File: lib/charms/data_platform_libs/v0/data_interfaces.py [hunk 1]
```diff
@@ -1426,6 +1441,196 @@ def _delete_relation_data(self, relation: Relation, fields: List[str]) -> None:
...
+        self, relation_name: str, relation_id: int, group_mapping: SecretGroup
+    ) -> str:
+        members = [self.charm.app.name]
+        if self.scope:
+            members.append(self.scope.value)
+        return f"{'.'.join(members)}"
+
+    def _generate_secret_field_name(self, group_mapping: SecretGroup = SecretGroup.EXTRA) -> str:
+        """Generate unique group_mappings for secrets within a relation context."""
+        return f"{self.secret_field_name}"
```
[the remark this answers is not in the document]
Comment (juditnovak) [reply lib/charms/data_platform_libs/v0/data_interfaces.py:1502]: No, it's not -- but I thought to keep consistency with the inherited signature.
Furthermore - the function WILL be called with a group mapping, so we definately need to allow for "extra args" here. Once we need to do that anyawy, I thought to keep it named + typed (as in parent).

File: lib/charms/data_platform_libs/v0/data_interfaces.py [hunk 2]
```diff
@@ -1426,6 +1441,196 @@ def _delete_relation_data(self, relation: Relation, fields: List[str]) -> None:
...
+            self.component, self.secret_fields, relation, fields
+        )
+
+    def _fetch_my_specific_relation_data(
+        self, relation: Relation, fields: Optional[List[str]]
+    ) -> Dict[str, str]:
+        """Fetch data available (directily or indirectly -- i.e. secrets) from the relation for owner/this_app."""
+        return self._fetch_relation_data_with_secrets(
+            self.component, self.secret_fields, relation, fields
+        )
```
[the remark this answers is not in the document]
Comment (juditnovak) [reply lib/charms/data_platform_libs/v0/data_interfaces.py:1578]: Almost ;-)

I assume you meant `Requires._fetch_specific_relation_data()` (as for `Provides` it's far too different, only checking the databag).

If you check  `Requires._fetch_specific_relation_data()` , you will find that the target is `relation.app`
I.e. "the other side".

While, in our case we are checking "our" side, i.e. `self.component` -- which may be our app or our unit.

The "nearly identical" function with the above is `DataProvides._fetch_my_specific_relation_data()` -- however that one needs to check the `Require` side's requrested secrets... So we can't just use it as it is....

File: lib/charms/data_platform_libs/v0/data_interfaces.py [hunk 3]
```diff
@@ -1426,6 +1441,196 @@ def _delete_relation_data(self, relation: Relation, fields: List[str]) -> None:
...
+    fetch_my_relation_field = leader_only(DataRelation.fetch_my_relation_field)
+
+
+# Base DataPeer
+
+
+class DataPeer(DataRequires, DataProvides):
+    """Represents peer relations."""
+
+    SECRET_FIELDS = ["admin", "monitor", "password", "tls", "tls-ca"]
```
[the remark this answers is not in the document]
Comment (juditnovak) [reply lib/charms/data_platform_libs/v0/data_interfaces.py:1456]: So... I wrote this from memory (I thought we had such fields in Mongo :sweat_smile: )

The intent was to define those fields that are somewhat "cross-charm standards" to be secrets. If we have any.

(I can remove `monitor` and `admin` and either replace with something else or just drop them completely)

File: lib/charms/data_platform_libs/v0/data_interfaces.py [hunk 4]
```diff
@@ -880,20 +881,20 @@ def _update_relation_data_without_secrets(
...
 
         for field in fields:
             try:
                 relation.data[app].pop(field)
             except KeyError:
-                logger.debug(
-                    "Non-existing field was attempted to be removed from the databag %s, %s",
-                    str(relation.id),
+                logger.error(
+                    "Non-existing secret '%s' was attempted to be removed from the databag (relation ID: %s)",
```
[the remark this answers is not in the document]
Comment (juditnovak) [reply lib/charms/data_platform_libs/v0/data_interfaces.py:895]: Indeed, I'll change that :-)

File: lib/charms/data_platform_libs/v0/data_interfaces.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (juditnovak) [reply lib/charms/data_platform_libs/v0/data_interfaces.py:1502]: If there's a more elegant solution here, I'm good to go with that :-)

File: lib/charms/data_platform_libs/v0/data_interfaces.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (marceloneppel) [reply lib/charms/data_platform_libs/v0/data_interfaces.py:1578]: I mean `DataPeer._fetch_specific_relation_data` and `DataPeer._fetch_my_specific_relation_data`.

[the remark this answers is not in the document]
Comment (juditnovak) [reply lib/charms/data_platform_libs/v0/data_interfaces.py:1578]: All right. That is a valid point.

Yes, I'm thinking either to disable or to "redefine" `fetch_realtion_data()` the same as `fetch_my_relation_data()` for Peer Relations.

I'm not decided on that.... As the two functionalities ARE redundant for Peers... 

What would you prefer? Disable or make them equal?

File: lib/charms/data_platform_libs/v0/data_interfaces.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (juditnovak) [reply lib/charms/data_platform_libs/v0/data_interfaces.py:1456]: Are there any "cross-charm" "reserved keywords" for secret data by any chance...?

[the remark this answers is not in the document]
Comment (marceloneppel) [reply lib/charms/data_platform_libs/v0/data_interfaces.py:1456]: AFAIK we don't such a list :/

File: lib/charms/data_platform_libs/v0/data_interfaces.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (marceloneppel) [reply lib/charms/data_platform_libs/v0/data_interfaces.py:1578]: I would keep `fetch_my_relation_field` and disable `fetch_relation_field`, as the latter is used (in non-peer relations) to get the other app's data, and here, we want the data from the current app.

**Review by delgod — APPROVED** [review]

File: lib/charms/data_platform_libs/v0/data_interfaces.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (juditnovak) [reply lib/charms/data_platform_libs/v0/data_interfaces.py]: OK. As a "balanced compromise", I leave `operator-password` here.

That's for sure used in Postgres and Mongo. And... If any other charm may wanna use it, it definately should be considered as a secret field :-)

File: lib/charms/data_platform_libs/v0/data_interfaces.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (juditnovak) [reply lib/charms/data_platform_libs/v0/data_interfaces.py:1577]: Done.

**Review by marceloneppel — APPROVED** [review]

**Review by delgod — APPROVED** [review]

**Pull request merged**